src/ModelSingleton.py

```diff
@@ -40,30 +40,119 @@ def sliding_window(self, text, window_size=512):
                 window_end = True
-                yield " ".join(current_string)
+                yield "".join(current_string)
+
+    def _pubannotate(self, q, inferred):
+        queries = [q.strip().split() for q in q]
+        ids_to_labels = {v: k for k, v in self.model._cfg.label_ids.items()}
+        start_idx = 0
+        end_idx = 0
+        denotations = []
+        for query in queries:
+            end_idx += len(query)
+            # extract predictions for the current query from the list of all predictions
+            preds = inferred[start_idx:end_idx]
+            start_idx = end_idx
+            for j, word in enumerate(query):
+                # strip out the punctuation to attach the entity tag to the word not to a punctuation mark
+                # that follows the word
+                if not word[-1].isalpha():
+                    word = word[:-1]
+                pad = 0 if j == 0 else 1
+                span_start = len(' '.join(query[:j])) + pad
+                span_end = span_start + len(word)
+
+                label = ids_to_labels[preds[j]]
+
+                is_not_pad_label = (label != self.model._cfg.dataset.pad_label and label != '0')
+
+                if not is_not_pad_label:
+                    # For things like fitness to practice where model labels it as fitness[B-biolink:NamedThing] to[0] practice[I-biolink:NamedThing]
+                    if j + 1 < len(query) and ids_to_labels[preds[j + 1]].startswith('I-'):
+                        denotations[-1]['text'] += " " + word
+                else:
+                    if label.startswith('I-'):
+                        denotations[-1]['span']['end'] = span_end
+                        denotations[-1]['text'] += " " + word
+                    else:
+                        label = label.replace('B-', '')
+                        denotation = {
+                            'id': f'I{j}-',
+                            'span': {
+                                'begin': span_start,
+                                'end': span_end
+                            },
+                            'obj': label,
+                            'text': word
+                        }
+                        denotations.append(denotation)
+        return {
+            'text': ''.join(q),
+            'denotations': denotations
+        }
+
+    def __add_predictions(
+            self, queries, batch_size: int = 32
+    ):
+        """
+        Add predicted token labels to the queries. Use this method for debugging and prototyping.
+        Args:
+            queries: text
+            batch_size: batch size to use during inference.
+            output_file: file to save models predictions
+        Returns:
+            result: text with added entities
+        """
+        inferred = self.model._infer(queries, batch_size)
+        return self._pubannotate(queries, inferred)
+
+    @staticmethod
+    def _merge_pubtator_annotations(annotations):
```

Comment:
I would change this to `pubannotator` or something like that to avoid confusion with the PubTator format.

```suggestion
    def _merge_pubannotator_annotations(annotations):
```

Reply:
(You'll also need to fix this function name elsewhere)

Reply:
fixed typo